src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Usage/ImplementGenericMathInterfacesCorrectly.cs

```diff
@@ -0,0 +1,146 @@
+                    break;
+                }
+            }
+
+            INamedTypeSymbol? baseType = symbol.BaseType;
+            while (baseType != null && baseType.IsGenericType)
+            {
+                foreach (INamedTypeSymbol anInterface in baseType.Interfaces)
+                {
+                    if (IsKnownInterfaceInTheChain(anInterface) &&
+                        FirstTypeParameterNameIsNotTheSymbolName(symbol, anInterface))
+                    {
+                        SyntaxNode? typeParameter = FindTheTypeArgumentOfTheInterfaceFromTypeDeclaration(symbol, symbol.BaseType);
+                        context.ReportDiagnostic(CreateDiagnostic(GMIRuleForClass, typeParameter, symbol.BaseType.OriginalDefinition.ToDisplayString(
+                            SymbolDisplayFormat.MinimallyQualifiedFormat), symbol.BaseType.OriginalDefinition.TypeParameters[0].Name, symbol));
+                        break;
+                    }
+                }
+
+                baseType = baseType.BaseType;
+            }
+
+            bool IsKnownInterfaceInTheChain(INamedTypeSymbol anInterface)
+            {
+                if (anInterface.IsGenericType)
+                {
+                    if (IsKnownInterface(anInterface, systemNS, systemNumericsNS))
+                    {
+                        return true;
+                    }
+                    else
+                    {
+                        foreach (INamedTypeSymbol parentInterface in anInterface.Interfaces)
+                        {
+                            if (IsKnownInterfaceInTheChain(parentInterface))
+                            {
+                                return true;
+                            }
+                        }
+                    }
+                }
+
+                return false;
+            }
+        }
+
+        private static Diagnostic CreateDiagnostic(DiagnosticDescriptor GMIRule, SyntaxNode? parameter,
+            string genericDeclaration, string parameterName, INamedTypeSymbol symbol) =>
+                parameter is null ? symbol.CreateDiagnostic(GMIRule, genericDeclaration, parameterName, symbol.Name) :
+                    parameter.CreateDiagnostic(GMIRule, genericDeclaration, parameterName, symbol.Name);
+
+        protected abstract SyntaxNode? FindTheTypeArgumentOfTheInterfaceFromTypeDeclaration(ISymbol typeSymbol, ISymbol theInterfaceSymbol);
+
+        private static bool IsKnownInterface(INamedTypeSymbol anInterface, INamespaceSymbol systemNS, INamespaceSymbol systemNumericsNS)
+        {
+            var iNamespace = anInterface.ContainingNamespace;
+
+            return s_knownInterfaces.Contains(anInterface.MetadataName) &&
+                   (iNamespace.Equals(systemNS, SymbolEqualityComparer.Default) ||
+                    iNamespace.Equals(systemNumericsNS, SymbolEqualityComparer.Default));
```

Comment:
It might be better if known interfaces are calculated as symbols in compilation start and do a symbol comparison instead of a string and a namespace symbol comparisons.

Reply:
I did not wanted to load all 34 interface in memory and loop through them each time, for me it is better to use its names in ImmutablArray of strings (I heard search from ImmutablArray is competent with hashtable), + checking namespaces would be same as checking all interfaces as symbols

Reply:
It doesn't seem `ImmutableArray` optimized for search, updated to `ImmutabeHashSet`